spark/src/main/scala/com/cloudera/livy/spark/client/ClientSession.scala

```diff
@@ -19,31 +19,46 @@ package com.cloudera.livy.spark.client
 import java.net.URI
 import java.nio.ByteBuffer
 import java.util.concurrent
-import java.util.concurrent.{ExecutionException, TimeUnit}
+import java.util.concurrent.TimeUnit
 import java.util.concurrent.atomic.AtomicLong
 
+import org.apache.spark.SparkConf
+
 import scala.concurrent.{ExecutionContext, Future}
 import scala.collection.JavaConverters._
 import scala.collection.mutable
 import scala.util.control.NonFatal
 
-import com.cloudera.livy.{Job, LivyClient}
+import com.cloudera.livy.{Logging, LivyClientBuilder}
 import com.cloudera.livy.client.local.LocalClient
 import com.cloudera.livy.sessions.{Session, SessionState}
 
-class ClientSession(val sessionId: Int, createRequest: CreateClientRequest) extends Session {
+class ClientSession(val sessionId: Int, createRequest: CreateClientRequest)
+  extends Session with Logging {
   implicit val executionContext = ExecutionContext.global
+
   var sessionState: SessionState = SessionState.Starting()
-  SessionClientTracker.createClient(
-    sessionId, createRequest.sparkConf.asJava, createRequest.timeout)
+
+  private val _timeout = createRequest.timeout
+
+  private val client = {
+    info("Creating LivyClient for sessionId: " + sessionId)
+    val builder = new LivyClientBuilder(new URI("local:spark"))
+    new SparkConf(true).getAll.foreach { case (k, v) => builder.setConf(k, v) }
+    builder
+      .setAll(createRequest.sparkConf.asJava)
+      .setConf("livy.client.sessionId", sessionId.toString)
+      .setIfMissing("spark.master", "yarn-cluster")
+      .build()
+  }
+  info("Livy client created.")
+
   sessionState = SessionState.Running()
 
   private val operations = mutable.Map[Long, java.util.concurrent.Future[_]]()
   private val operationCounter = new AtomicLong(0)
 
-  def getClient(): Option[LocalClient] = {
-    SessionClientTracker.getClient(id).map(_.asInstanceOf[LocalClient])
-  }
+  @volatile private var _lastActivity = System.currentTimeMillis()
```

Comment:
Good point. nano resolution would allow to use this for some sort of profiling too. It's not possible / not accurate enough to do with ms resolution.